### ISSUE TITLE
[🐛] Firestore Batch is hanging on set/update

### ISSUE THREAD

**Issue by RayVR** [invertase/react-native-firebase#5929]
## Issue

Setting/updating documents using batch is failing. Some of the use cases are somewhat difficult to explain and I have not recreated this outside of my app, however switching from batch to individual document set/update has resolved the hanging issues.

here's an example, which may or may not recreate the issue. I've tested this on macOS and linux and both iOS and Android.
```js
const parent = firestore().doc("collectionName/docName")
d1 = firestore().doc("collectionName/docName/subcoll/1")
d2 = firestore().doc("collectionName/docName/subcoll/2")
d3 = firestore().doc("collectionName/docName/subcoll/3")

batch = firestore().batch()

batch.set(d1, {'a': 123})
batch.set(d2, {'a': 123})
batch.set(d3, {'a': 123})

const updater = () => {
return parent.update({info: "new info"}).then(() => {
 return batch.commit()
})
}
```
---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "Braced",
  "version": "6.2.0",
  "private": true,
  "scripts": {
    "start": "react-native start",
    "run:android": "react-native run-android",
    "run:ios": "react-native run-ios --simulator=\"iPhone X\"",
    "build:apk": "cd android && ./gradlew assembleRelease",
    "test": "jest",
    "prepare": "patch-package"
  },
  "dependencies": {
    "@eva-design/eva": "^2.1.1",
    "@expo/react-native-action-sheet": "^3.12.0",
    "@flyerhq/react-native-android-uri-path": "^2.2.2",
    "@flyerhq/react-native-chat-ui": "^1.4.3",
    "@flyerhq/react-native-firebase-chat-core": "^1.4.2",
    "@motify/skeleton": "^0.16.0",
    "@notifee/react-native": "^3.0.2",
    "@react-native-async-storage/async-storage": "^1.15.5",
    "@react-native-community/blur": "^3.6.0",
    "@react-native-community/cameraroll": "^4.0.4",
    "@react-native-community/hooks": "^2.8.1",
    "@react-native-community/slider": "^4.1.12",
    "@react-native-firebase/analytics": "13.1.0",
    "@react-native-firebase/app": "13.1.0",
    "@react-native-firebase/auth": "13.1.0",
    "@react-native-firebase/crashlytics": "^13.1.0",
    "@react-native-firebase/firestore": "13.1.0",
    "@react-native-firebase/functions": "^13.1.0",
    "@react-native-firebase/messaging": "^13.1.0",
    "@react-native-firebase/storage": "^13.1.0",
    "@react-navigation/bottom-tabs": "^6.0.5",
    "@react-navigation/native": "^6.0.2",
    "@react-navigation/native-stack": "^6.2.5",
    "@react-navigation/stack": "^6.0.7",
    "@types/json-stable-stringify": "^1.0.32",
    "@types/lodash": "^4.14.176",
    "@types/node": "^16.11.2",
    "@types/object-hash": "^2.1.0",
    "@types/uuid": "^8.3.3",
    "@types/webpack-env": "^1.16.0",
    "@ui-kitten/components": "^5.0.0",
    "@ui-kitten/eva-icons": "^5.1.1",
    "convert": "1.9.1",
    "dayjs": "^1.10.6",
    "expo": "^43.0.0",
    "expo-av": "^10.1.2",
    "expo-keep-awake": "^10.0.0",
    "expo-linear-gradient": "^10.0.3",
    "expo-splash-screen": "~0.13.5",
    "expo-status-bar": "^1.1.0",
    "faker": "^5.5.3",
    "formik": "^2.2.6",
    "fuse.js": "^6.4.6",
    "json-stable-stringify": "^1.0.1",
    "moti": "^0.16.0",
    "object-hash": "^2.1.1",
    "react": "17.0.2",
    "react-error-boundary": "^3.1.3",
    "react-native": "0.66.3",
    "react-native-console-time-polyfill": "^1.2.3",
    "react-native-document-picker": "^7.1.1",
    "react-native-draggable-flatlist": "^2.5.4",
    "react-native-elements": "^1.2.7",
    "react-native-fast-image": "^8.3.7",
    "react-native-file-viewer": "^2.1.4",
    "react-native-gesture-handler": "^1.10.3",
    "react-native-get-random-values": "^1.7.1",
    "react-native-gifted-chat": "^0.16.3",
    "react-native-image-picker": "^4.0.4",
    "react-native-keyboard-aware-scroll-view": "^0.9.5",
    "react-native-pressable-opacity": "^1.0.4",
    "react-native-progress": "^4.1.2",
    "react-native-reanimated": "2.2.4",
    "react-native-redash": "^16.2.3",
    "react-native-safe-area-context": "^3.2.0",
    "react-native-screens": "3.9.0",
    "react-native-static-safe-area-insets": "^2.1.1",
    "react-native-storage": "^1.0.1",
    "react-native-svg": "12.1.1",
    "react-native-switch-selector": "^2.0.3",
    "react-native-vector-icons": "8.1.0",
    "react-native-vision-camera": "^2.9.4",
    "react-native-webview": "^11.6.5",
    "reanimated-bottom-sheet": "^1.0.0-alpha.22",
    "recoil": "0.5.2",
    "util": "^0.12.1",
    "uuid": "^8.3.2",
    "validator": "^12.2.0",
    "yup": "^0.32.9"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9",
    "@babel/runtime": "^7.12.9",
    "@react-native-community/eslint-config": "^2.0.0",
    "@types/jest": "^25.1.1",
    "@types/react": "^17.0.31",
    "@types/react-native": "0.65.10",
    "@types/react-test-renderer": "^16.9.2",
    "@types/validator": "^12.0.1",
    "babel-jest": "^26.6.3",
    "eslint": "^8.0.1",
    "eslint-plugin-react-hooks": "^4.2.0",
    "jest": "^26",
    "jest-mock": "^26",
    "metro-react-native-babel-preset": "^0.66.2",
    "patch-package": "^6.1.4",
    "react-addons-test-utils": "^15.6.2",
    "react-native-typescript-transformer": "^1.2.13",
    "react-test-renderer": "17.0.2",
    "ts-jest": "^26",
    "typescript": "^4.3.5"
  },
  "resolutions": {
    "react-native/@jest/create-cache-key-function": "^27.0.2"
  },
  "jest": {
    "preset": "react-native",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "transform": {
      "^.+\\.(js)$": "<rootDir>/node_modules/babel-jest",
      "\\.(ts|tsx)$": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/__tests__/data/"
    ],
    "transformIgnorePatterns": [
      "node_modules/(?!((jest-)?@?react-native|react-navigation|@react-navigation/.*|@react-native-community/masked-view))"
    ],
    "cacheDirectory": ".jest/cache",
    "setupFiles": [
      "./node_modules/react-native-gesture-handler/jestSetup.js"
    ]
  }
}
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
require File.join(File.dirname(`node --print "require.resolve('expo/package.json')"`), "scripts/autolinking")
require File.join(File.dirname(`node --print "require.resolve('react-native/package.json')"`), "scripts/react_native_pods")
require File.join(File.dirname(`node --print "require.resolve('@react-native-community/cli-platform-ios/package.json')"`), "native_modules")


platform :ios, '12.0'

# To override the React Native Firebase iOS SDK versions used uncomment any of the below and change the version
# $FirebaseSDKVersion = '6.8.1'
# $FabricSDKVersion = '1.6.0'
# $CrashlyticsSDKVersion = '3.1.0'

ENV['SWIFT_VERSION'] = '5.2'
require 'json'
podfile_properties = JSON.parse(File.read('./Podfile.properties.json')) rescue {}


target 'Braced' do
  use_expo_modules!
  # Pods for Braced
  pod 'Firebase/Analytics'
  pod 'Firebase/Auth'
  pod 'Firebase/Firestore'

  config = use_native_modules!
  use_react_native!(
    :path => config[:reactNativePath],
    # to enable hermes on iOS, change `false` to `true` and then install pods
    :hermes_enabled => true
  )

	
  pod 'RNVectorIcons', :path => '../node_modules/react-native-vector-icons'
  target 'BracedTests' do
    inherit! :search_paths
    # Pods for testing
  end

  post_install do |installer|
    react_native_post_install(installer)

    # Workaround `Cycle inside FBReactNativeSpec` error for react-native 0.64
    # Reference: https://github.com/software-mansion/react-native-screens/issues/842#issuecomment-812543933
    installer.pods_project.targets.each do |target|
      if (target.name&.eql?('FBReactNativeSpec'))
        target.build_phases.each do |build_phase|
          if (build_phase.respond_to?(:name) && build_phase.name.eql?('[CP-User] Generate Specs'))
            target.build_phases.move(build_phase, 0)
          end
        end
      end
    end
    __apply_Xcode_12_5_M1_post_install_workaround(installer)
  end


end


$NotifeeExtension = true

target 'BracedNotificationService' do
  pod 'RNNotifeeCore', :path => '../node_modules/@notifee/react-native/RNNotifeeCore.podspec'
end
```

#### `AppDelegate.m`:

```objc
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "AppDelegate.h"
#import "RNFBMessagingModule.h"
#import <Firebase.h>

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import <React/RCTLinkingManager.h>
#import <React/RCTConvert.h>





@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{

  if ([FIRApp defaultApp] == nil) {
    [FIRApp configure];
  }

  NSDictionary *appProperties = [RNFBMessagingModule addCustomPropsToUserProps:nil withLaunchOptions:launchOptions];

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                             moduleName:@"Braced"
                                             initialProperties:appProperties];

  id rootViewBackgroundColor = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"RCTRootViewBackgroundColor"];
  if (rootViewBackgroundColor != nil) {
    rootView.backgroundColor = [RCTConvert UIColor:rootViewBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [super application:application didFinishLaunchingWithOptions:launchOptions];
  UIStoryboard *sb = [UIStoryboard storyboardWithName:@"LaunchScreen" bundle:nil];
  UIViewController *vc = [sb instantiateInitialViewController];
  rootView.loadingView = vc.view;

  return YES;
}

- (NSArray<id<RCTBridgeModule>> *)extraModulesForBridge:(RCTBridge *)bridge
{
    // If you'd like to export some custom RCTBridgeModules that are not Expo modules, add them here!
    return @[];

}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  return [RCTLinkingManager application:application openURL:url options:options];
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(nonnull NSUserActivity *)userActivity
 restorationHandler:(nonnull void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler
{
 return [RCTLinkingManager application:application
                  continueUserActivity:userActivity
                    restorationHandler:restorationHandler];
}


@end
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [x] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: macOS 12.0.1
    CPU: (4) x64 Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz
    Memory: 168.98 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 17.2.0 - /usr/local/bin/node
    Yarn: 1.21.1 - /usr/local/bin/yarn
    npm: 8.1.4 - /usr/local/bin/npm
    Watchman: 2021.12.06.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.2 - /Users/rroberts/.gem/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
    Android SDK: Not Found
  IDEs:
    Android Studio: 2020.3 AI-203.7717.56.2031.7583922
    Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_292 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: Not Found
    react-native: Not Found
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `13.1.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `Firestore`
- **Are you using `TypeScript`?**
  - `Y` & `4.3.5`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
Hmmm :thinking: this smells like an upstream issue. I'm not aware (though I could be mistaken...) of anything we're doing in our wrapper layer that would cause problems here, so I think this will be in the underlying SDKs.
You don't mention if it is platform specific or both android and ios
Have you tried it with firebase-js-sdk? Just out of curiosity
Have you tried a quickstart reproduction on the affect (if it is just one platform) SDK, or your favorite, if it is both?
Additionally - if you do have a possible use case (and it appears you do) - you might translate that into an e2e test so we actually have a reproduction - https://github.com/invertase/react-native-firebase/blob/main/packages/firestore/e2e/issues.e2e.js

Wish I had some ambient awareness of something that was more on target, but this looks like a "needs a lot more info and some initial investigation at multiple layers..." type issue

**Comment by RayVR** [comment]
@mikehardy Thanks for the quick response. 
This is happening on both platforms. 

I have not tested with the js sdk yet. I will have more time to investigate later this week. Until then I wanted to raise this here in case others hit an issue with the batch code. 

Oddly, a very complex batch of deletes had no issues but everywhere I use set/update with a batch it fails.

**Comment by RayVR** [comment]
@mikehardy I'm looking into this today.

After a full night of sleep I'm less certain the issue is with firestore batch. I do see multiple writes being created and only one makes it through.

I will share more information soon.

**Comment by RayVR** [comment]
I have not been able to figure out why this is happening but it's not related to the batch. I installed several older versions of react-native-firebase and they all exhibited the same behavior so that eliminates my initial concern that this was related to the upgrade to 13.1.0.

At this point the most likely cause is a timing issue in my own code.  

If it turns out to be related to firebase in any way I'll create a new ticket with what I've learned. Apologies for the noise.
